docs/standard/io/pipelines.md

```diff
@@ -0,0 +1,342 @@
+
+* Returns an incomplete `ValueTask<FlushResult>` when the amount of data in the `Pipe` crosses `PauseWriterThreshold`.
+* Completes `ValueTask<FlushResult>` when it becomes lower than `ResumeWriterThreshold`.
+
+Two values are used to prevent rapid cycling, which can occur if one value is used.
+
+### Examples
+
+```csharp
+// The Pipe will start returning incomplete tasks from FlushAsync until
+// the reader examines at least 5 bytes.
+var options = new PipeOptions(pauseWriterThreshold: 10, resumeWriterThreshold: 5);
+var pipe = new Pipe(options);
+```
+
+### PipeScheduler
+
+Typically when using `async` and `await`, asynchronous code resumes on either on a <xref:System.Threading.Tasks.TaskScheduler> or on the current  <xref:System.Threading.SynchronizationContext>.
+
+When doing I/O, it's important to have fine-grained control over where the I/O is performed. This control allows taking advantage of CPU caches effectively. Efficient caching is critical for high-performance apps like web servers. <xref:System.IO.Pipelines.PipeScheduler> provides control over where asynchronous callbacks run. By default:
+
+* The current <xref:System.Threading.SynchronizationContext> is used.
+* If there's no `SynchronizationContext`, it uses the thread pool to run callbacks.
+
+[!code-csharp[](~/samples/snippets/csharp/pipelines/Program.cs?name=snippet)]
+
+### Pipe reset
+
+It's frequently efficient to reuse the `Pipe` object. To reset the pipe, call <xref:System.IO.Pipelines.PipeReader> <xref:System.IO.Pipelines.Pipe.Reset%2A> when both the `PipeReader` and `PipeWriter` are complete.
+
+## PipeReader
+
+<xref:System.IO.Pipelines.PipeReader> manages memory on the caller's behalf. **Always** call <xref:System.IO.Pipelines.PipeReader.AdvanceTo%2A?displayProperty=nameWithType> after calling <xref:System.IO.Pipelines.PipeReader.ReadAsync%2A?displayProperty=nameWithType>. This lets the `PipeReader` know when the caller is done with the memory so that it can be tracked. The `ReadOnlySequence<byte>` returned from `PipeReader.ReadAsync` is only valid until the call the `PipeReader.AdvanceTo`:
+
+* It's illegal to use `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo`.
+* Using `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo` is undefined.
+
+`PipeReader.AdvanceTo` takes two <xref:System.SequencePosition> arguments:
+
+* The first argument determines how much memory was consumed.
+* The second argument determines how much of the buffer was observed.
+
+Marking data as consumed means that the pipe can return the memory to the underlying buffer pool. Marking data as observed controls what the next call to `PipeReader.ReadAsync` does. Marking everything as observed means that the next call to `PipeReader.ReadAsync` won't return until there's more data written to the pipe. Any other value will make the next call to `PipeReader.ReadAsync` return immediately with the unobserved data.
```

Comment:
```suggestion
Marking data as consumed means that the pipe can return the memory to the underlying buffer pool. Marking data as observed controls what the next call to `PipeReader.ReadAsync` does. Marking everything as observed means that the next call to `PipeReader.ReadAsync` won't return until there's more data written to the pipe. Any other value will make the next call to `PipeReader.ReadAsync` return immediately with the observed *and* unobserved data, but data that has already been consumed.
```

I know the point of the last sentence is to communicate that the next call to ReadAsync will always return immediately of you didn't observe everything in the last call, but se should make it very clear that ReadResult includes data that has already been observed. To this point in the documentation, I don't think it's that clear.

Reply:
Feel free to use the suggested edit feature 😬

Reply:
I thought I did. Didn't I?

Reply:
> I thought I did. Didn't I?

Yes, I took it in a batch.